### DISCUSSION TITLE
is there any C++ grpc server emebeded in the milvus system?

### DISCUSSION THREAD

**Discussion by caozhuozi** [milvus-io/milvus#34577] [Q&A and General discussion] [answered]
I recently found an error which should be throwed from C++ code (after searching the error text `write index to fd error, errorCode is`). 
I know there is a `HandleCStatus` function that logs all the C errors in GO (as long as the C code is called in Go, any error should be logged, right?). But I cannot find any `CStatus returns err` logs.

How could this heppen?

im wondering if there is any C++ server in the system such that its C code is not handled by Go?

milvus: `2.3.5`

@yhmo Could you please help answer this?

**Comment by yhmo** [comment]
This is a case that a CStatus is thrown:
https://github.com/milvus-io/milvus/issues/34088

[2024/06/24 09:46:44.539 +08:00] [WARN] [segments/cgo_util.go:54] ["CStatus returns err"] [error=std::exception]

**Reply by caozhuozi** [reply]
The weird thing is that we could NOT find any `CStatus returns err` logs but only sth like this:
```
[cluster/worker.go:75] ["failed to call LoadSegments, worker return error"] [traceID=XXX] [workerID=XXX] [errorCode=UnexpectedError] [reason="write index to fd error, errorCode is -1: service internal error: UnknownError"]
```

**Reply by caozhuozi** [reply]
I think any C error should be caught by `HandleCStatus` function and should be loggged. But I could NOT find any `CStatus returns err` in our logs.

**Comment by yhmo** [comment]
There are two HandleCStatus functions:
https://github.com/milvus-io/milvus/blob/9aceff5a6e34eefc747cfbce07dde7d2f7db859e/internal/util/initcore/init_core.go#L118
https://github.com/milvus-io/milvus/blob/9aceff5a6e34eefc747cfbce07dde7d2f7db859e/internal/querynodev2/segments/cgo_util.go#L43

Search "C Runtime Exception:" in your log.

**Reply by caozhuozi** [reply]
Still cannot find `C Runtime Exception` logs. Could you confirm: In milvus, all the C errors are expected to be caught by these **2** `HandleCStatus`, right?  

**Are there other situations where C errors occur but are not caught by these **2** `HandleCStatus`**?

**Reply by yhmo** [reply]
I cannot confirm because the source code is written by 50+ engineers. You have to read the source code by your self.

**Reply by caozhuozi** [reply]
Hi, @yhmo 😂  I know there would be some potential bugs in Milvus. My original intention behind my question was actually this: Are all the C errors **expected to be** caught by these two `HandleCStatus` functions (or **theoretically**, **by design**)? If so, I would focus on reading the code with the intention of finding bugs.

**Reply by yhmo** [reply, marked as answer]
I suppose all the C errors should be caught by the two functions. But maybe someone handles a C error by different code, I am not sure.

**Reply by caozhuozi** [reply]
oh sry its our bad... we lost some logs thats the reason we could not find the `CStatus returns error` logs...
